common/archiver/s3store/historyArchiver.go

```diff
@@ -168,29 +173,86 @@ func (h *historyArchiver) Archive(
-		return err
-	}
+		key := constructHistoryKey(URI.Path(), request.DomainID, request.WorkflowID, request.RunID, request.CloseFailoverVersion, progress.BatchIdx)
 
-	key := constructHistoryKey(URI.Path(), request.DomainID, request.WorkflowID, request.RunID, request.CloseFailoverVersion)
+		exists, err := keyExists(ctx, h.s3cli, URI, key)
+		if err != nil {
+			logger := logger.WithTags(tag.ArchivalArchiveFailReason(errWriteKey), tag.Error(err))
+			if isRetryableError(err) {
+				logger.Error(archiver.ArchiveTransientErrorMsg)
+			} else {
+				logger.Error(archiver.ArchiveNonRetriableErrorMsg)
+			}
+			return err
+		}
+		blobSize := int64(binary.Size(encodedHistoryBlob))
+		if exists {
+			scope.IncCounter(metrics.HistoryArchiverBlobExistsCount)
+		} else {
+			if err := upload(ctx, h.s3cli, URI, key, encodedHistoryBlob); err != nil {
+				logger := logger.WithTags(tag.ArchivalArchiveFailReason(errWriteKey), tag.Error(err))
+				if isRetryableError(err) {
+					logger.Error(archiver.ArchiveTransientErrorMsg)
+				} else {
+					logger.Error(archiver.ArchiveNonRetriableErrorMsg)
+				}
+				return err
+			}
+			progress.uploadedSize += blobSize
+			scope.RecordTimer(metrics.HistoryArchiverBlobSize, time.Duration(blobSize))
+		}
 
-	if err := upload(ctx, h.s3cli, URI, key, encodedHistoryBatches); err != nil {
-		logger.Error(archiver.ArchiveNonRetriableErrorMsg, tag.ArchivalArchiveFailReason(errWriteKey), tag.Error(err))
-		return err
+		progress.historySize += blobSize
+		progress.BatchIdx = progress.BatchIdx + 1
+		saveHistoryIteratorState(ctx, featureCatalog, historyIterator, &progress)
 	}
-	totalUploadSize := int64(binary.Size(encodedHistoryBatches))
-	scope.AddCounter(metrics.HistoryArchiverTotalUploadSize, totalUploadSize)
-	scope.AddCounter(metrics.HistoryArchiverHistorySize, totalUploadSize)
-	scope.IncCounter(metrics.HistoryArchiverArchiveSuccessCount)
 
+	scope.RecordTimer(metrics.HistoryArchiverTotalUploadSize, time.Duration(progress.uploadedSize))
+	scope.RecordTimer(metrics.HistoryArchiverHistorySize, time.Duration(progress.historySize))
+	scope.IncCounter(metrics.HistoryArchiverArchiveSuccessCount)
 	return nil
 }
 
+func loadHistoryIterator(ctx context.Context, request *archiver.ArchiveHistoryRequest, historyManager persistence.HistoryManager, featureCatalog *archiver.ArchiveFeatureCatalog, progress *uploadProgress) (historyIterator archiver.HistoryIterator) {
+	if featureCatalog.ProgressManager != nil {
+		if featureCatalog.ProgressManager.HasProgress(ctx) {
+			err := featureCatalog.ProgressManager.LoadProgress(ctx, progress)
+			if err == nil {
+				historyIterator, err := archiver.NewHistoryIteratorFromState(request, historyManager, targetHistoryBlobSize, progress.IteratorState)
+				if err == nil {
+					return historyIterator
+				}
+			}
+			progress.IteratorState = nil
+			progress.BatchIdx = 0
+			progress.historySize = 0
+			progress.uploadedSize = 0
```

Comment:
Those four lines can be remove right?

Reply:
Well if NewHistoryIteratorFromState fails we need to start from the beginning. Otherwise BatchIdx might be a different number for instance

Reply:
I was also unsure if LoadProgress could potentially modify progress and fail

Reply:
Makes sense, you are totally right. I had some misunderstanding. Sorry.